docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc

```diff
@@ -1078,6 +1078,112 @@ public @interface CanWrite {
+
+The `@PermissionsAllowed` annotation attribute `params` can refer to methods or fields of the secured method parameter.
+You can use this feature to pass `jakarta.ws.rs.BeanParam` parameters into your custom permission.
+Consider following `jakarta.ws.rs.BeanParam` class:
+
+[source,java]
+----
+package org.acme.crud;
+
+import java.util.List;
+
+import jakarta.ws.rs.HeaderParam;
+import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.SecurityContext;
+import jakarta.ws.rs.core.UriInfo;
+
+public class SimpleBeanParam {
+
+    @QueryParam("query")
+    private String privateQuery;
+
+    @HeaderParam("Authorization")
+    String authorizationHeader;
+
+    @Context
+    SecurityContext securityContext;
+
+    @Context
+    UriInfo uriInfo;
+
+    String getPrivateQuery() {
+        return privateQuery;
+    }
+}
+----
+
+And let's say we have `SimplePermission` where we need to check `Authorization` header and principal name.
+That can be done like this:
+
+[source,java]
+----
+package org.acme.crud;
+
+import java.security.Permission;
+import java.util.Objects;
+
+public class SimplePermission extends Permission {
+
+    private final String authorizationHeader;
+    private final String principalName;
+
+    public SimpleBeanParamPermission(String permissionName, String authorizationHeader, String name, String privateQuery) { <1> <2>
+        super(permissionName);
+        this.authorizationHeader = authorizationHeader;
+        this.principalName = name;
+        if ("private-query".equals(privateQuery)) {
+            // your business logic
+        }
+    }
+
+    @Override
+    public boolean implies(Permission requiredPermission) {
+        if (requiredPermission instanceof SimpleBeanParamPermission that) {
+            return "MyAuthorization".equals(that.authorizationHeader)
+                && "admin".equals(that.principalName);
+        }
+        return false;
+    }
+
+    ...
+}
+----
+<1> Constructor parameters `authorizationHeader` and `principalName` corresponds to `SimpleBeanParam` fields with same names.
+<2> Constructor parameter `privateQuery` is accessed with a getter method.
+Quarkus always prefer method access, therefore, first we look for the `privateQuery` method, then for the `getPrivateQuery` method.
+If neither method can be found on the `SimpleBeanParam` class, Quarkus tries to find the `privateQuery` field.
+
+The `SimplePermission` can be used on your Jakarta REST resource like in the example below:
+
+[source,java]
+----
+package org.acme.crud;
+
+import io.quarkus.security.PermissionsAllowed;
+import jakarta.ws.rs.BeanParam;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+
+@Path("/hello")
+public class SimpleResource {
+
+    @PermissionsAllowed(value = "read", permission = SimplePermission.class,
+        params = { "param.authorizationHeader", "param.privateQuery", "param.securityContext.getUserPrincipal.name" }) <1>
```

Comment:
I've been wondering, why can't a custom permission just accept a `SimpleBeanParam` parameter ?

Reply:
Noone stops you. But in actual application you will have different DTOs and you want to check just one field. Why should you create separate permission accepting different DTOs or introduce level of abstraction (like interface) instead of `param.authorizationHeader`?

I don't say people should use it for multiple levels of nested `@BeanParam`s , but I don't believe that your proposal is simpler than writin `param.authorizationHeader` and having `Permission` that accepts just `authorizationHeader`. It's clear what this permission does.

I've added this explanation to the issue description based on your question. Thank you

Reply:
BTW I am sure your question will be also Stephane argument, so that might be where it ends anyway... 

it's just matter of opinion. Please note that user that opened that issue considered ability to pass just `id` there convenient: https://github.com/quarkusio/quarkus/issues/43231#issuecomment-2348263322

Reply:
I'm sorry, I'm not quite sure what are the problems with passing a bean param directly. Are you saying it already works ? Then one why support expressions like `beanParam.a`, can't just a custom permission decide itself what it wants from it ? It is just a container for several parameters in one place...

Reply:
> I'm sorry, I'm not quite sure what are the problems with passing a bean param directly. Are you saying it already works ?

Method parameter annotated with a `@BeanParam` is just another method parameter. It works.

> Then why support expressions like beanParam.a can't just a custom permission decide itself what it wants from it ?

User expected that he can pass `id` from `PermissionParam` and it didn't work. IIUC he considered that convenient and so do I. I don't see why should all the endpoints consume same `PermissionParam` bean parameter just because we don't allow to pass one `String` there. This way, you can have bean parameters based on what you actually need and not having there `PermissionParam` repeated everywhere.

> It is just a container for several parameters in one place...

It's just a container repeated on every endpoint if you need to pass just `id`, some endpoints will be `@GET` and will have `id` as path parameter, some endpoints will have `id` inside payload DTO as attribute. And this way, you have one permission accepting `id`, not more.

Reply:
>I don't see why should all the endpoints consume same PermissionParam bean parameter just because we don't allow to pass one String there.

IMHO it is unnecessary, this is actually a variation of the non-typed mini expression language which we are somewhat concerned about. A user has to type both the bean name and the property inside that name, 3 times im the doc-ed example. Quarkus has to do some code to support all that, and for what, to save users doing `bean.getSomeParam()` in the custom permission class ?

Reply:
The whole concept of `BeanParam` is let uses avoid type individual parameters in the JAX-RS method, it should be consistent if this bean is accessed elsewhere

Reply:
@michalvavrik I'm really sorry, I saw the issue earlier but I was under an impression `objects` can be passed as parameters, I don't know why... What if someone registers a `ParamConverter` to convert a JAX-RS `String` query param to `MyString` bean, do we have to then support unwrapping that String for custom permissions too ? IMHO it is not worth it

Reply:
I don't know how else to explain this. I gave you arguments why and you don't share them. As far as I am concerned, being able to have one permission for many data types by simply typing `param.nested` is great and reduce repetition. This implementation takes care of this during the build time, no extra code.

Let's not continue as I don't have better arguments than I have already provided. Thanks for your time.

Reply:
So, this appears to be deconstruction, where the many possible places where `@PermissionAllowed` is allowed will outnumber the single definition of the `Permission` constructor it binds to.

I suppose it is more reusable if the annotated methods vary in how they obtain their parameters (bean params, directly) but I fear that in most cases, they will obtain the parameters in the same way and so there will be a lot of duplication of `params = { "param.authorizationHeader", "param.privateQuery", "param.securityContext.getUserPrincipal.name" }` unless they define a meta-annotation, in which case it's fine.